service_discovery/core/src/registry_impl.rs

```diff
@@ -0,0 +1,457 @@
+                DiscoverByNamespaceResponse { services: service_list };
+            Ok(Response::new(discover_by_namespace_response))
+        }
+    }
+
+    /// Discovers a single service based on its "fully qualified name", consisting of the namespace,
+    /// name, and version of the service.
+    ///
+    /// This function discovers a service based on a [`DiscoverServiceRequest`]. Returns a
+    /// [`DiscoverServiceResponse`].
+    ///
+    /// # Arguments
+    ///
+    /// * `request` - A [`DiscoverServiceRequest`] wrapped by a [`tonic::Request`].
+    async fn discover_service(
+        &self,
+        request: Request<DiscoverServiceRequest>,
+    ) -> Result<Response<DiscoverServiceResponse>, Status> {
+        let request_inner = request.into_inner();
+
+        let service_identifiers = ServiceIdentifiers {
+            namespace: request_inner.namespace.clone(),
+            name: request_inner.name.clone(),
+            version: request_inner.version,
+        };
+
+        // This block controls the lifetime of the lock.
+        {
+            let lock: RwLockReadGuard<HashMap<ServiceIdentifiers, ServiceMetadata>> =
+                self.registry_map.read();
+            match lock.get(&service_identifiers) {
+                Some(service) => {
+                    info!("Read service in DiscoverService {:?}", service);
+                    let discover_service_response =
+                        DiscoverServiceResponse { service: Some(service.clone()) };
+                    Ok(Response::new(discover_service_response))
+                }
+                None => {
+                    let not_found_message = format!(
+                        "No service found for namespace: {0}, name: {1}, version: {2}",
+                        service_identifiers.namespace,
+                        service_identifiers.name,
+                        service_identifiers.version
+                    );
+                    warn!(not_found_message);
+                    Err(Status::not_found(not_found_message))
+                }
+            }
+        }
+    }
+
+    /// Inspects the contents of the service registry.
+    ///
+    /// This function retrieves all services currently registered based on an [`InspectRequest`]. Returns a
+    /// [`InspectResponse`].
+    ///
+    /// # Arguments
+    ///
+    /// * `request` - A [`InspectRequest`] wrapped by a [`tonic::Request`].
+    async fn inspect(
+        &self,
+        _request: Request<InspectRequest>,
+    ) -> Result<Response<InspectResponse>, Status> {
+        let lock: RwLockReadGuard<HashMap<ServiceIdentifiers, ServiceMetadata>> =
+            self.registry_map.read();
+        let services_list = lock.values().cloned().collect();
+        let inspect_response = InspectResponse { services: services_list };
+        Ok(Response::new(inspect_response))
+    }
+}
+
+/// Identifiers for a given service.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct ServiceIdentifiers {
```

Comment:
we calls this the plural, but it represents a single service; what about calling it the singular?

Reply:
updated. I was thinking of it as there are multiple "identifiers" per service (i.e. name, version, etc.) but I think "ServiceIdentifier" can be clearer